internal/tool/malicious_packages_scanner.go

```diff
@@ -0,0 +1,238 @@
+package tool
+
+import (
+	"compress/gzip"
+	"encoding/json"
+	"fmt"
+	"os"
+	"slices"
+	"strings"
+
+	ptypes "github.com/aquasecurity/trivy/pkg/types"
+	codacy "github.com/codacy/codacy-engine-golang-seed/v6"
+	"github.com/samber/lo"
+	"golang.org/x/mod/semver"
+)
+
+// MaliciousPackagesIndexPath is the default path to the malicious package index.
+const MaliciousPackagesIndexPath = "/dist/cache/codacy-trivy/openssf-malicious-packages-index.json.gz"
+
+// maliciousPackage represents a shallow representation of an Open Source Vulnerability (OSV).
+// Although it's schema is generic, it is guaranteed that it is only instantiated for Malicious Package vulnerabilities.
+//
+// See https://ossf.github.io/osv-schema/
+type maliciousPackage struct {
+	// OpenSSF identifier of the malicious package.
+	ID string `json:"id"`
+	// A summary of why the package is malicious.
+	Summary string `json:"summary"`
+	// The versions of the malicious package.
+	// The version syntax is the one defined by the package ecosystem where the malicious package is deployed.
+	Versions []string `json:"versions"`
+	// The range of versions considered malicious.
+	// This is usually defined if `Versions` is empty, but sometimes both are defined.
+	Ranges []maliciousPackageRange `json:"ranges"`
+}
+
+// matchesVersion checks if the reported malicious package versions match version.
+//
+// `Ranges` is only checked if there is no direct match in `Versions`.
+func (o maliciousPackage) matchesVersion(version string) bool {
+	if slices.Contains(o.Versions, version) {
+		return true
+	}
+	for _, affectedRange := range o.Ranges {
+		if affectedRange.matchesVersion(version) {
+			return true
+		}
+	}
+	return false
+}
+
+// maliciousPackageRange represents range of versions considered malicious.
+//
+// See https://ossf.github.io/osv-schema/#affectedranges-field
+type maliciousPackageRange struct {
+	Type   string                       `json:"type"`
+	Events []maliciousPackageRangeEvent `json:"events"`
+}
+
+// matchesVersion checks if version matches any of the range events but only if range is of type '[SEMVER]'.
+//
+// [SEMVER]: https://ossf.github.io/osv-schema/#affectedrangestype-field
+func (r maliciousPackageRange) matchesVersion(version string) bool {
+	if r.Type != "SEMVER" {
+		return false
+	}
```

Comment:
Non-`SEMVER` ranges need specific logic for proper version matching which we don't do in this PR.

Looking at OSV schema, we're missing support for `GIT` and `ECOSYSTEM` ranges. In practive, OSSF data does not have any `GIT` ranges but it has plenty of `ECOSYSTEM` ranges.

Let's discuss what we can do about that but maybe in another PR.